libcudacxx/include/cuda/std/__string/constexpr_c_functions.h

```diff
@@ -0,0 +1,349 @@
+#define _LIBCUDACXX___STRING_CONSTEXPR_C_FUNCTIONS_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cstddef/types.h>
+#include <cuda/std/__type_traits/is_constant_evaluated.h>
+#include <cuda/std/__type_traits/make_nbit_int.h>
+#include <cuda/std/__type_traits/remove_const.h>
+#include <cuda/std/climits>
+
+#if !_CCCL_COMPILER(NVRTC)
+#  include <cstring>
+#endif // !_CCCL_COMPILER(NVRTC)
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <class _CharT>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr _CharT*
+__cccl_constexpr_strcpy(_CharT* _CCCL_RESTRICT __dst, const _CharT* _CCCL_RESTRICT __src) noexcept
+{
+  if constexpr (sizeof(_CharT) == 1)
+  {
+    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
+    {
+      NV_IF_TARGET(NV_IS_HOST,
+                   (return reinterpret_cast<_CharT*>(
+                             ::strcpy(reinterpret_cast<char*>(__dst), reinterpret_cast<const char*>(__src)));))
+    }
+  }
+
+  _CharT* __dst_it = __dst;
+  while ((*__dst_it++ = *__src++) != _CharT('\0'))
+  {
+  }
+  return __dst;
+}
+
+template <class _CharT>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr _CharT*
+__cccl_constexpr_strncpy(_CharT* _CCCL_RESTRICT __dst, const _CharT* _CCCL_RESTRICT __src, size_t __n) noexcept
+{
+  if constexpr (sizeof(_CharT) == 1)
+  {
+    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
+    {
+      NV_IF_TARGET(NV_IS_HOST,
+                   (return reinterpret_cast<_CharT*>(
+                             ::strncpy(reinterpret_cast<char*>(__dst), reinterpret_cast<const char*>(__src), __n));))
+    }
+  }
+
+  _CharT* __dst_it = __dst;
+  while (__n--)
+  {
+    if ((*__dst_it++ = *__src++) == _CharT('\0'))
+    {
+      while (__n--)
+      {
+        *__dst_it++ = _CharT('\0');
+      }
+      break;
+    }
+  }
+  return __dst;
+}
+
+template <class _CharT>
+[[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr size_t __cccl_constexpr_strlen(const _CharT* __ptr) noexcept
+{
+  if constexpr (sizeof(_CharT) == 1)
+  {
+    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
+    {
+      NV_IF_TARGET(NV_IS_HOST, (return ::strlen(reinterpret_cast<const char*>(__ptr));))
+    }
+  }
+
+  size_t __len = 0;
+  while (*__ptr++ != _CharT('\0'))
+  {
+    ++__len;
+  }
+  return __len;
+}
+
+template <class _CharT>
+[[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr int
+__cccl_constexpr_strcmp(const _CharT* __lhs, const _CharT* __rhs) noexcept
+{
+  if constexpr (sizeof(_CharT) == 1)
+  {
+    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
+    {
+      NV_IF_TARGET(NV_IS_HOST,
+                   (return ::strcmp(reinterpret_cast<const char*>(__lhs), reinterpret_cast<const char*>(__rhs));))
```

Comment:
We should be able to expand all of that to use `__builtin_memcmp` 

That would avoid the cstring include and also generalize to larger types